DOC: Enhance asfreq docs

plammens opened the pull request:
- [n/a] closes #xxxxx
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] tests added

Enhanced some docstrings for time-series related methods; mainly `asfreq`. Namely:
 - Converted `NDFrame.asfreq` into a template to use the specific `klass` in `Series.asfreq` and `DataFrame.asfreq`
 - Fully explain the behaviour of `NDFrame.asfreq` in the extended summary.
 - Add an extended summary to `asfreq` in `PeriodArray`/`PeriodIndex`
 - Use `klass` kwarg in docstring template for `PeriodAray.asfreq`\`PeriodIndex.asfreq`
 - Add an extended summary to `date_range`
 - Other very minor fixes/enhancements


The original motivator for this PR was [this question](https://stackoverflow.com/q/63860589/6117426), in particular the confusion of the asker.

plammens committed:
DOC: enhance documentation for asfreq

plammens committed:
DOC: enhance documentation for date_range

Added extended summary.

plammens committed:
DOC: fix cross-reference to period_array in PeriodArray

MarcoGorelli commented on pandas/core/arrays/period.py:
```diff
@@ -489,15 +493,19 @@ def _time_shift(self, periods, freq=None):
         freq : str
             A frequency.
-        how : str {'E', 'S'}
+        how : str {{'E', 'S'}}, default 'E'
```
Thanks @plammens - I haven't tried running this yet, but why are you doubling the brackets here? Single brackets seems consistent with the docstring guide https://pandas.pydata.org/docs/development/contributing_docstring.html

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -7527,26 +7528,39 @@ def asfreq(
+
+        If the index of this {klass} is a :class:`~pandas.PeriodIndex`, the new index
+        is the result of transforming the original index with
+        :meth:`PeriodIndex.asfreq pandas.PeriodIndex.asfreq` (so the original index
```
Did you mean to write it twice here?

plammens replied:
Hmm, I don't see it, what do you mean?

MarcoGorelli replied:
```
:meth:`PeriodIndex.asfreq pandas.PeriodIndex.asfreq`
```

Is it correct to have `PeriodIndex.asfreq` twice?

plammens replied:
Oh I see! That's Sphinx syntax, the first part is the link text (`PeriodIndex.asfreq`) and the second part is the actual link path (`pandas.PeriodIndex.asfreq`). I did this just to omit the leading `pandas.` from the link text. (Here `~` wouldn't do it because it would just show `asfreq`.)

MarcoGorelli replied:
I'm probably missing something, but when I build this locally it shows up twice rather than having a link:
![image](https://user-images.githubusercontent.com/33491632/93300610-03219b80-f7ef-11ea-8db2-89c757c222a8.png)

MarcoGorelli replied:
Should it be something like
```suggestion
        :meth:`PeriodIndex.asfreq <pandas.PeriodIndex.asfreq>` (so the original index
```
?

plammens replied:
Ah, you're totally correct! For some reason I interpreted the angle brackets as not part of the syntax itself—and I didn't notice that when checking the output.

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -7527,26 +7528,39 @@ def asfreq(
+
+        Otherwise, the new index will be equivalent to ``pd.date_range(start, end,
+        freq=freq)`` where ``start`` and ``end`` are, respectively, the first and
+        last entries in the original index. (see :func:`pandas.date_range`). The
```
No full-stop needed here IMO
```suggestion
        last entries in the original index (see :func:`pandas.date_range`). The
```

plammens committed:
CLN: Apply suggestions from code review

Co-authored-by: Marco Gorelli <m.e.gorelli@gmail.com>

plammens commented:
@MarcoGorelli  Alright, it should be all done now. I've extracted a generic `call_once` context manager (for the warnings thing) to `pandas.util._call_once`, and added tests for that. The two of the CI builds are failing on a completely unrelated test (`tests.plotting.TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works`), and I can't for the life of me figure out why.

MarcoGorelli commented:
> The two of the CI builds are failing on a completely unrelated test (`tests.plotting.TestDataFrameGroupByPlots.test_series_groupby_plotting_nominally_works`), and I can't for the life of me figure out why.

They're almost certainly unrelated, don't worry about them

Will take another look later, thanks for updating!

MarcoGorelli commented:
Does this have to all be part of the same PR, or could `check_deprecated_resample_kwargs` be a separate one?

plammens commented:
> Does this have to all be part of the same PR, or could `check_deprecated_resample_kwargs` be a separate one?

It has to be part of the same PR because of how the warnings were managed previously: the warnings were raised in `Grouper.__new__`, for all possible paths that led there, and the `stacklevel` was set based on whether the call came from a `TimeGrouper` or not (because at the time, every public call of `resample` ended up at `TimeGrouper` in exactly three stack frames).

https://github.com/pandas-dev/pandas/blob/8df0218a4bfae50a7191ff0869e63c28884fb733/pandas/core/groupby/grouper.py#L227-L268

This is very brittle—as this PR demonstrates. When I made the changes to `resample` docs, I made `NDFrame.resample` into a template, to allow for customized `DataFrame.resample` and `Series.resample` docs. This introduces a new stack frame since these methods have to call `super().resample(...)`: 

https://github.com/pandas-dev/pandas/blob/7e8812318d62278712999f6b8351a237f3a6c582/pandas/core/frame.py#L9127-L9156

This then breaks the hacky `stacklevel` munging mentioned above since now for `DataFrame` and `Series` the `stacklevel` should technically have been `5`.

In fact if you run the tests on commit 27b16cf, you will see that they fail on `pandas.tests.resample.test_deprecated`. When I added the `check_resample_deprecated_kwargs` thing in e78a415 to fix this issue.

MarcoGorelli commented:
Thanks @plammens for explaining - from what I've understood, you want to share docstrings between `DataFrame.resample` and `Series.resample`. However, this causes problems with how the warnings are managed in `Grouper.__new__`.

Would it be possible to make one PR to change how warnings are managed in `Grouper.__new__`? You write

> This then breaks the hacky stacklevel munging mentioned above

so would it be possible to have 1 PR to improve the stacklevel munging and then, once that's in, another PR to share docstrings between `DataFrame.resample` and `Series.resample`? Sorry if I've misunderstood, in which case please do correct me

plammens commented:
> Thanks @plammens for explaining - from what I've understood, you want to share docstrings between `DataFrame.resample` and `Series.resample`. However, this causes problems with how the warnings are managed in `Grouper.__new__`.

Yes, exactly.

I've made a new PR (#36629), but until that gets merged, if I remove the relevant commits from here, this PR will be failing the tests, so should I wait until the other PR gets merged?

MarcoGorelli commented:
> I've made a new PR (#36629),

Awesome, thanks

> so should I wait until the other PR gets merged?

Yup

plammens commented:
> This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.

Waiting for #36629 (or an equivalent fix) to be merged.

jreback commented:
> > This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.
> 
> Waiting for #36629 (or an equivalent fix) to be merged.

why is this a blocker?

MarcoGorelli commented:
> > > This pull request is stale because it has been open for thirty days with no activity. Please update or respond to this comment if you're still interested in working on this.
> > 
> > 
> > Waiting for #36629 (or an equivalent fix) to be merged.
> 
> why is this a blocker?

As far as I understand, it's because he wants to share docs between `DataFrame.resample` and `Series.resample`, though doing so would change the stacklevel

jreback commented:
let just fix the appropriate tests (eg turn off the stack level checks)

these aren't that important for a really old warning

MarcoGorelli commented:
Sure - happy to do that, @plammens ? Sorry for having led you down the wrong path

plammens commented:
> let just fix the appropriate tests (eg turn off the stack level checks)
> 
> these aren't that important for a really old warning

There is already an open PR for that - #37603 :)

plammens committed:
REF: Extract resmple warning code into helper function

plammens committed:
BUG: Update stacklevel to match refactoring

See #37603 for rationale regarding this solution and the removal of some
of the warning tests.

plammens commented:
After discussion in #37603, it was decided to just update the stacklevel accordingly and disable some of the stacklevel checks for those cases in which the stacklevel couldn't be computed easily (at the same time as the one for `Series.resample` and `DataFrame.resample`). I updated the PR to reflect this.

I rebased onto the latest passing master. There have been lots of changes meanwhile, so it is possible I messed up something while resolving merge conflicts.

CC @jorisvandenbossche

jreback reviewed (approved):
lgtm @jbrockmendel @jorisvandenbossche if comments

jbrockmendel commented on pandas/core/arrays/period.py:
```diff
@@ -67,7 +71,7 @@ class PeriodArray(PeriodMixin, dtl.DatelikeOps):
 
-    Users should use :func:`period_range` to create new instances.
+    Users should use :func:`~pandas.period_array` to create new instances.
     Alternatively, :func:`array` can be used to create new instances
```
should "array" be changed to the same pattern as for period_array?

plammens replied:
I assume this is `numpy.array`?

jbrockmendel replied:
im pretty sure this refers to pd.array

plammens replied:
Changed in 9c91ee7.

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -7717,26 +7717,39 @@ def asfreq(
+        freq=freq)`` where ``start`` and ``end`` are, respectively, the first and
+        last entries in the original index (see :func:`pandas.date_range`). The
+        values corresponding to any timesteps in the new index which were not present
+        in the original index will be null (``NaN``), unless a method for filling
```
NaT?

plammens replied:
I'm not sure what you mean, the behaviour as shown in the examples uses `NaN`?

jbrockmendel replied:
yep, my bad

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -7764,7 +7777,7 @@ def asfreq(
         >>> index = pd.date_range('1/1/2000', periods=4, freq='T')
         >>> series = pd.Series([0.0, None, 2.0, 3.0], index=index)
-        >>> df = pd.DataFrame({'s':series})
+        >>> df = pd.DataFrame({{'s':series}})
```
`'s':series` -> `"s": series`?

plammens replied:
Thanks, I'll add the space; but all of the other examples use single quotes, so I'd stick to that.

plammens replied:
Fixed in 9c91ee7.

jbrockmendel commented on pandas/core/generic.py:
```diff
@@ -8214,8 +8229,8 @@ def resample(
-        >>> d = {'price': [10, 11, 9, 13, 14, 18, 17, 19],
-        ...      'volume': [50, 60, 40, 100, 50, 100, 40, 50]}
+        >>> d = {{'price': [10, 11, 9, 13, 14, 18, 17, 19],
+        ...     'volume': [50, 60, 40, 100, 50, 100, 40, 50]}}
```
why was the space removed here?

plammens replied:
Probably a typo, thanks.

plammens replied:
Fixed in 9c91ee7.

jbrockmendel commented on pandas/core/indexes/datetimes.py:
```diff
@@ -930,6 +930,15 @@ def date_range(
     """
     Return a fixed frequency DatetimeIndex.
 
+    Returns the range of equally spaced timesteps (where the time offset between any
```
timesteps -> timestamps?

jbrockmendel replied:
time offset -> difference?

plammens replied:
I used "timesteps" here to refer to each of the time points in the time series; maybe this is not standard terminology though I think it's used elsewhere in the docs. Maybe should use "time points"?

plammens replied:
Edited in 9c91ee7, does it look good now?

plammens committed:
DOC: Apply suggestions from code review

Co-authored-by: jbrockmendel <jbrockmendel@gmail.com>

jreback commented:
thanks @plammens very nice

jreback commented:
pls have a look at the built docs (may take a bit): https://pandas.pydata.org/pandas-docs/dev/ and if needed followup
